src/include/migraphx/iterator.hpp

```diff
@@ -203,7 +215,16 @@ struct iterator_operators
     template <class I, class U = T>
     constexpr auto operator[](I n) const -> decltype(*(static_cast<const U&>(*this) + n))
     {
-        return *(static_cast<const U&>(*this) + n);
+        auto it = static_cast<const U&>(*this) + n;
+        // cppcheck-suppress migraphx-ConditionalAssert
```

Comment:
Why suppress this or does this always flag an error with the assert?

Reply:
There isnt a way currently in a cppcheck addon to check for a ` if constexpr`  vs just `if`, unfortunately. Normally when you write `if(b) { assert(c); }` it should be written as `assert(b && c)`, but in this case its a `if constexpr` condition to ensure the `assert` is correct at compile-time because if `b` is false then `c` wont compile.

Reply:
Yeah lets discuss this. This seems like maybe we should make this not an assert but a stronger condition we always check which would drop this suppress but also indicate an error on failure since we don't want that behaviour to be occurring. Caller/user can then error and move on/retry if the reference is incorrect.